cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,380 @@
+            if (originalBrokerBrokerSetId == null) {
+                throw new OptimizationFailureException(String.format("[%s] Cannot find broker set for broker id %s", name(), replica.originalBroker()
+                        .id()));
+            }
+
+            // If the re-balancing action is performed within the broker set boundary, we are good
+            if (broker.isAlive() && originalBrokerBrokerSetId.equals(currentBrokerSetId)) {
+                continue;
+            }
+            // If the brokerSet awareness condition is violated. Move replica to an eligible broker
+            Set<Broker> eligibleBrokers = brokersByBrokerSet.get(originalBrokerBrokerSetId)
+                    .stream()
+                    .filter(b -> clusterModel.aliveBrokers()
+                            .contains(b))
+                    .collect(Collectors.toSet());
+            if (maybeApplyBalancingAction(clusterModel, replica, eligibleBrokers,
+                                          ActionType.INTER_BROKER_REPLICA_MOVEMENT, optimizedGoals, optimizationOptions) == null) {
+                // If balancing action can not be applied, provide recommendation to add new Brokers.
+                ProvisionRecommendation recommendation = new ProvisionRecommendation.Builder(ProvisionStatus.UNDER_PROVISIONED)
+                        .numBrokers(clusterModel.maxReplicationFactor())
+                        .build();
+                throw new OptimizationFailureException(String.format("[%s] Cannot move replica %s to %s on brokerSet %s",
+                                                                     name(), replica, eligibleBrokers,
+                                                                     originalBrokerBrokerSetId), recommendation);
+            }
+        }
+    }
+
+    /**
+     * Check whether the given action is acceptable by this goal. The following actions are acceptable:
+     * <ul>
+     *   <li>All leadership moves</li>
+     *   <li>Replica moves that do not violate {@link #doesReplicaMoveViolateActionAcceptance(Replica, Broker)}</li>
+     *   <li>Swaps that do not violate {@link #doesReplicaMoveViolateActionAcceptance(Replica, Broker)}
+     *   in both direction</li>
+     * </ul>
+     *
+     * @param action       Action to be checked for acceptance.
+     * @param clusterModel The state of the cluster.
+     * @return {@link ActionAcceptance#ACCEPT} if the action is acceptable by this goal,
+     * {@link ActionAcceptance#BROKER_REJECT} if the action is rejected due to violating brokerSet awareness in the destination
+     * broker after moving source replica to destination broker, {@link ActionAcceptance#REPLICA_REJECT} otherwise.
+     */
+    @Override
+    public ActionAcceptance actionAcceptance(BalancingAction action, ClusterModel clusterModel) {
+        switch (action.balancingAction()) {
+            case LEADERSHIP_MOVEMENT:
```

Comment:
From the goal implementation, isn't the only possible action for this goal is "INTER_BROKER_REPLICA_MOVEMENT"? (line 284,285)? I don't see any other possible actions from the current goal implementation.

Reply:
My understanding is that this is called by other goals in AnalyzerUtils.isProposalAcceptableForOptimizedGoals() and other goals can propose LEADERSHIP_MOVEMENT. BrokerSetAwareGoal does not have any say in leadership movement - since leadership movements occurs within existing replicas.

Reply:
If the above is correct than can this comment be resolved ?